cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -479,11 +479,25 @@ Future<std::shared_ptr<parquet::arrow::FileReader>> ParquetFileFormat::GetReader
                                                          default_fragment_scan_options));
   auto properties =
       MakeReaderProperties(*this, parquet_scan_options.get(), options->pool);
-  ARROW_ASSIGN_OR_RAISE(auto input, source.Open());
+  auto input_fut = source.OpenAsync();
   // TODO(ARROW-12259): workaround since we have Future<(move-only type)>
-  auto reader_fut = parquet::ParquetFileReader::OpenAsync(
-      std::move(input), std::move(properties), metadata);
+
   auto path = source.path();
+  auto reader_fut = input_fut.Then(
+    [=](const std::shared_ptr<arrow::io::RandomAccessFile>&) mutable
+      -> Result<std::unique_ptr<parquet::ParquetFileReader>> {
+        ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::io::RandomAccessFile> input,
+                              input_fut.MoveResult());
+        auto rfut = parquet::ParquetFileReader::OpenAsync(std::move(input), std::move(properties), metadata);
+        ARROW_ASSIGN_OR_RAISE(auto reader,
+                              rfut.MoveResult());
+        return reader;
+      },
+      [path](
+          const Status& status) -> Result<std::unique_ptr<parquet::ParquetFileReader>> {
+        return WrapSourceError(status, path);
+      });
```

Comment:
Would `WrapSourceError` called twice since L516-L519 already called that?

Reply:
Good point, I removed the callback